sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts

```diff
@@ -0,0 +1,146 @@
+async function processError(err: Error, sessionId?: string) {
+  console.log(`[${sessionId}] had error ${err.message}`);
+}
+
+// Called if we are closing a session
+// `reason` will be:
+// * 'error' if we are closing because of an error(the error will be delivered
+//   to `processError` above)
+// * 'idle_timeout' if we `sessionIdleTimeoutMs` milliseconds pass without
+//   any messages being received (ie, session can be considered empty).
+//
+async function processClose(reason: "error" | "idle_timeout", sessionId: string) {
+  if (reason === "error") {
+    console.log(`[${sessionId}] was closed because of error`);
+  } else if (reason === "idle_timeout") {
+    console.log(`[${sessionId}] was closed - no more messages within idle timeout`);
+  }
+}
+
+// utility function to create a timer that can be refreshed
+function createRefreshableTimer(timeoutMs: number, resolve: Function): () => void {
+  let timer: any;
+
+  return () => {
+    clearTimeout(timer);
+    timer = setTimeout(() => resolve(), timeoutMs);
+  };
+}
+
+// Queries Service Bus for the next available session and processes it.
+async function receiveFromNextSession(serviceBusClient: ServiceBusClient): Promise<void> {
+  let sessionReceiver: SessionReceiver<ReceivedMessageWithLock>;
+
+  try {
+    sessionReceiver = await serviceBusClient.createSessionReceiver(queueName, "peekLock", {
+      autoRenewLockDurationInMs: sessionIdleTimeoutMs
+    });
+  } catch (err) {
+    if ((err as MessagingError).code === "SessionCannotBeLockedError") {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
+      return;
+    }
+
+    await processError(err, undefined);
+    return;
+  }
+
+  await processInitialize(sessionReceiver.sessionId);
+
+  const sessionFullyRead = new Promise((resolveSessionAsFullyRead, rejectSessionWithError) => {
+    const refreshTimer = createRefreshableTimer(sessionIdleTimeoutMs, resolveSessionAsFullyRead);
+    refreshTimer();
+
+    sessionReceiver.subscribe(
+      {
+        async processMessage(msg) {
+          refreshTimer();
+          await processMessage(msg);
+        },
+        async processError(err) {
+          rejectSessionWithError(err);
+        }
+      },
+      {
+        abortSignal: abortController.signal
+      }
+    );
+  });
+
+  try {
+    await sessionFullyRead;
+    await processClose("idle_timeout", sessionReceiver.sessionId);
+  } catch (err) {
+    await processError(err, sessionReceiver.sessionId);
+    await processClose("error", sessionReceiver.sessionId);
+  } finally {
+    await sessionReceiver.close();
+  }
+}
+
+async function roundRobinThroughAvailableSessions(abortSignal: AbortSignalLike): Promise<void> {
+  const serviceBusClient = new ServiceBusClient(serviceBusConnectionString);
+
+  for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
+    (async () => {
+      while (!abortSignal.aborted) {
```

Comment:
When the abort signal is fired, we stop spinning any more sessions, but what about already created session receivers, how would they stop?

Reply:
subscribe() is also passed the same abortSignal.

Reply:
Ah! subscribe is getting the signal directly from the controller. I would either expect the signal to be passed to `receiveFromNextSession()` or have `roundRobinThroughAvailableSessions()` access it directly from the controller too.

Reply:
Ah, I meant to use the global abortController just to simplify things.

Fixed (by removing the parameters altogether)